compiler/qsc_data_structures/src/index_map.rs

```diff
@@ -14,6 +14,23 @@ pub struct IndexMap<K, V> {
     values: Vec<Option<V>>,
 }
 
+impl<K, V> IndexMap<K, V>
+where
+    K: Into<usize>,
+    V: Default,
+{
+    pub fn get_mut_or_default(&mut self, key: K) -> &mut V {
+        let index: usize = key.into();
+        if index >= self.values.len() {
+            self.values.resize_with(index + 1, Default::default);
```

Comment:
I think we want `None` here not `Default`. If there are gaps in the vector before the index you're inserting, you want them initialized with `None` to indicate there's nothing at that index. Otherwise you're implicitly initializing the gaps.

```rust
let i: IndexMap<usize, usize> = IndexMap::default(); // no elements in the index map

let e = i.get_mut_or_default(33);  // only idx 33 should be initialized

i.contains_key(22); // should be false

for (k,v) in i {
  // should yield only one element
}
```

Reply:
You could just call `self.insert(...)` to be safe.

Reply:
AAAHHH I was definitely reading this as initializing this to `Some(T::default())`. 

Nah, I wouldn't say it's unclear, just a reading mistake on my part, it happens.